### ISSUE TITLE
"conch" command line apparently completely broken again

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3809]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)| @glyph reported|
|-|-|
|Trac ID|trac#3809|
|Type|release blocker: regression|
|Created|2009-04-30 22:05:59Z|
|Branch|https://github.com/twisted/twisted/tree/conch-client-broken-again-3809|

This regression appears to have been introduced by r26534.  The test coverage introduced in that revision is apparently insufficient.

```
$ conch twistedmatrix.com
/home/glyph/Projects/Twisted/trunk/twisted/conch/ssh/userauth.py:660: DeprecationWarning: Returning a string from SSHUserAuthClient.getPublicKey() is deprecated since Twisted 9.0.  Return a keys.Key() instead.
  DeprecationWarning)
Traceback (most recent call last):
  File "/home/glyph/Projects/Twisted/trunk/twisted/internet/selectreactor.py", line 146, in _doReadOrWrite
    why = getattr(selectable, method)()
  File "/home/glyph/Projects/Twisted/trunk/twisted/internet/tcp.py", line 460, in doRead
    return self.protocol.dataReceived(data)
  File "/home/glyph/Projects/Twisted/trunk/twisted/conch/ssh/transport.py", line 313, in dataReceived
    self.dispatchMessage(messageNum, packet[1:])
  File "/home/glyph/Projects/Twisted/trunk/twisted/conch/ssh/transport.py", line 335, in dispatchMessage
    messageNum, payload)
--- <exception caught here> ---
  File "/home/glyph/Projects/Twisted/trunk/twisted/python/log.py", line 84, in callWithLogger
    return callWithContext({"system": lp}, func, *args, **kw)
  File "/home/glyph/Projects/Twisted/trunk/twisted/python/log.py", line 69, in callWithContext
    return context.call({ILogContext: newCtx}, func, *args, **kw)
  File "/home/glyph/Projects/Twisted/trunk/twisted/python/context.py", line 59, in callWithContext
    return self.currentContext().callWithContext(ctx, func, *args, **kw)
  File "/home/glyph/Projects/Twisted/trunk/twisted/python/context.py", line 37, in callWithContext
    return func(*args,**kw)
  File "/home/glyph/Projects/Twisted/trunk/twisted/conch/ssh/service.py", line 44, in packetReceived
    return f(packet)
  File "/home/glyph/Projects/Twisted/trunk/twisted/conch/ssh/userauth.py", line 533, in ssh_USERAUTH_PK_OK
    return func(packet)
  File "/home/glyph/Projects/Twisted/trunk/twisted/conch/ssh/userauth.py", line 547, in ssh_USERAUTH_PK_OK_publickey
    d  = self.signData(publicKey, b)
  File "/home/glyph/Projects/Twisted/trunk/twisted/conch/client/default.py", line 193, in signData
    return self.keyAgent.signData(publicKey, signData)
  File "/home/glyph/Projects/Twisted/trunk/twisted/conch/ssh/agent.py", line 102, in signData
    req = NS(blob)
  File "/home/glyph/Projects/Twisted/trunk/twisted/conch/ssh/common.py", line 42, in NS
    return struct.pack('!L',len(t)) + t
exceptions.TypeError: object of type 'Key' has no len()
Connection to twistedmatrix.com closed.
```

Please fix ASAP; I am not reverting the branch as there are now conflicts in trunk, but I'll go to the trouble of resolving them if this isn't a trivial fix.

<details><summary>Searchable metadata</summary>

```
trac-id__3809 3809
type__release_blocker__regression release blocker: regression
reporter__glyph glyph
priority__highest highest
milestone__Twisted_9_0 Twisted-9.0
branch__branches_conch_client_broken_again_3809 branches/conch-client-broken-again-3809
branch_author__z3p__therve__exarkun z3p, therve, exarkun
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__ 
time__1241129159000000 1241129159000000
changetime__1241380221000000 1241380221000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="z3p's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|z3p commented|
|-|-|

(In [26778]) Branching to 'conch-client-broken-again-3809'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve removed owner|
|-|-|

I've added a test in the attached branch. It seems to work fine now.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph set owner to @exarkun|
|-|-|

  1. The tests don't skip properly.  Try `trial --without-module Crypto.Cipher.DES3 twisted.conch.test.test_default`; this is how e.g. Jython will see a pycrypto install.  Please make the skip contingent upon that, the same as `test_checkers`.
    1. File a ticket for refactoring the without-pycrypto-skip behavior so that we don't have to address this separately everywhere.
  1. I think the test is sufficient for fixing the regression, but this really needs a better test.
    1. The word "correctly" should never be used in a test docstring.  Explain what's happening and what constitutes a "correct" behavior, don't just say "correct".
    1. Feel free to file a separate ticket for this, but: that blob of binary data in the test is lame.  Either something should parse the data so we can make structured assertions (such that the signature verifies) or we should be talking to a verified fake of `KeyAgent`.
  1. `twisted.conch.client.default.SSHUserAuthClient.signData` has no docstring, and the new behavior apparently disagrees with the docstring of its superclass documentation, `twisted.conch.ssh.userauth.SSHUserAuthClient`, which documents its `@type` as `C{str}`.  `twisted.conch.userauth.SSHUserAuthClient.lastPublicKey` is _also_ documented to be a `str`, so something else has gone wrong that `ssh_USERAUTH_PK_OK_publickey` is passing a `Key` object in.  Based on reading what's in the branch, the problem is in `_cbGetPublicKey`, which expects its `publicKey` argument to be a `Key` and turns it into one if it's given a `str`.  I leave it to you to decide how to actually fix this mess.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun removed owner|
|-|-|

  1. Changed how the tests get skipped.  [#7308](https://github.com/twisted/twisted/issues/7308) is a ticket which is related to how this code should all be refactored to reduce the duplication.  This will involve making all of Conch at least importable, though, so that the tests can stop pulling all the `ImportError`-handling tricks they do now.
  1. 
    1. Fixed
    1. Until the interfaces are better specified, I think it makes sense for this to be an integration test, so I didn't make a fake of `KeyAgent` (although I started to because that would be pretty nice).  So, I removed most of the binary blobbiness.
  1. The docstring for `lastPublicKey` and `usedPublicKeys` was not in line with the implementation.  The keys are stored as `Key` instances, not blobs.  I corrected the documentation.

All fixes in r26803.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph set owner to @exarkun|
|-|-|


Looking good.  I like what you did with the symbolic names in the test; it's a lot easier for me to read now.  Please fix the following minor issues:

    1. I don't see that anything was changed about the way that the tests are being skipped.  It's still just importing `PyCrypto` and catching the `ImportError`.  Please actually fix that before merging.  (Heck, it was _your_ idea! :))
    1. `twisted.conch.client.default.SSHUserAuthClient.signData`'s docstring now has a `@type` annotation that says "data" is a `str`, but actually the parameter's name is `signData`.  The docstring should match the implementation.
    1. `twisted.conch.ssh.userauth.SSHUserAuthClient.signData`'s docstring:
        1. still says that `publicKey` is a `str`; it should say it's a `Key`
        1. contains some trailing whitespace; can you remove it as long as you're editing it.
    1. This stuff could be deferred to a separate ticket, because it's nominally not part of this change, but it's also both trivial and related, so do these at your discretion:
        1. `twisted.conch.ssh.userauth.SSHUserAuthClient.getPublicKey` still documents itself as returning a `str` even though `_cbGetPublicKey` will emit a deprecation warning if you actually return that.  It also doesn't explain that you need to override it in subclasses if it's actually supposed to do anything, and what will happen if subclasses return a `str`.
        1. `twisted.conch.client.default.SSHUserAuthClient.getPublicKey` is undocumented, but appears to be providing the deprecated behavior.  It would be good if it at least described its behavior; there should also be a separate ticket for fixing it to return a key object rather than a string.
        1. It would be nice to add some VWS around the various methods that were modified to make them all separated by 2 blank linse.

Then go ahead and merge.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)|@therve set owner to @therve|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve set owner to @exarkun|
|-|-|

I've handle the first 3 points.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [26811]) Update the documentation for `SSHUserAuthClient.getPublicKey

refs [#3809](https://github.com/twisted/twisted/issues/3809)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [26812]) Give the agent-aware `SSHUserAuthClient` subclass's `getPublicKey` method a docstring

refs [#3809](https://github.com/twisted/twisted/issues/3809)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [26813]) add extra whitespace between methods which were changed in this branch anyway

refs [#3809](https://github.com/twisted/twisted/issues/3809)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

(In [26814]) fix epytext markup typo

refs [#3809](https://github.com/twisted/twisted/issues/3809)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

4.1 and 4.3 fixed in r26812/r26814 and r26813.  4.2 somewhat addressed, but mostly I expect it to be dealt with in the newly filed [#3813](https://github.com/twisted/twisted/issues/3813).

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/conch-client-broken-again-3809) look good, so merging.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun set status to `closed`|
|-|-|

(In [26815]) Merge conch-client-broken-again-3809

Author: z3p, therve, exarkun
Reviewer: glyph
Fixes: [#3809](https://github.com/twisted/twisted/issues/3809)

Fix the mis-use of the SSH agent client by `twisted.conch.client.default`:
the agent's `signData` method expects the key as a string, not an object.
Update the documentation for related APIs which was either missing or which
documented the types of various key-related APIs incorrectly.

This allows the command-line conch client to use an ssh agent successfully
again.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
